### PR TITLE
libobs: Add wayland hotkey infrastructure

### PR DESCRIPTION
**Pull request by kkartaltepe**
### Description

Users on wayland are displeased that they cannot see their hotkey
bindings. This enables keys reported like X11, and has the infrastructure
in place in case wayland ever decides to allow for capturing input.

Also fix numpad subtract missing on X11.

### Motivation and Context

Users cannot even see or set hotkeys which might be usable while OBS is in focus. Hotkeys now work while OBS is in focus and display properly on the UI.

### How Has This Been Tested?

Tested in kde that keymapping tables work correctly on US layouts and the keys on my keyboard that work in X11 work in Wayland.

More testing on non-US layouts would be appreciated.

### Types of changes

 - Bug fix (non-breaking change which fixes an issue) 

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
Also most of the key translation was done programmatically, some keys may be missing the only ones I added manually were numpad keys.

File: libobs/obs-nix-x11.c [hunk 1]
```diff
@@ -1179,6 +1179,8 @@ static void obs_nix_x11_key_to_str(obs_key_t key, struct dstr *dstr)
 		return translate_key(key, "Numpad *");
 	case OBS_KEY_NUMPLUS:
 		return translate_key(key, "Numpad +");
+	case OBS_KEY_NUMMINUS:
+		return translate_key(key, "Numpad -");
```
Comment (WizardCM) [inline libobs/obs-nix-x11.c:1183]: This should likely be a separate commit.

Comment (kkartaltepe) [reply libobs/obs-nix-x11.c:1183]: Sounds good.

File: libobs/obs-nix-wayland.c [hunk 2]
```diff
@@ -19,7 +19,224 @@
...
+{
+	obs_hotkeys_platform_t *plat = (obs_hotkeys_platform_t *)data;
+	if (key >= MAX_KEYCODES)
+		return;
+
+	const xkb_keysym_t *syms;
+	// My setup doesnt have multiple layouts, do people actually use these?
+	// int num_layouts = xkb_keymap_num_layouts_for_key(keymap, key);
+	// blog(LOG_INFO, "key layouts: %d", num_layouts);
+	for (int l = 0; l < MAX_SHIFT_LEVELS; l++) {
```
Comment (GeorgesStavracas) [inline libobs/obs-nix-wayland.c:58]: `int level` perhaps?

File: libobs/obs-nix-wayland.c [hunk 3]
```diff
@@ -19,7 +19,224 @@
...
+#include <sys/mman.h> // mmap
+#include <string.h>   // strcmp
+
 #include <wayland-client.h>
+#include <xkbcommon/xkbcommon.h>
+
+// X11 only supports 256 scancodes, most keyboards dont have 256 keys so this should be reasonable.
+#define MAX_KEYCODES 256
+// X11 keymaps only have 4 shift levels, im not sure xkbcommon supports a way to shift the state into a higher level anyway.
+#define MAX_SHIFT_LEVELS 4
```
Comment (GeorgesStavracas) [inline libobs/obs-nix-wayland.c:32]: Can this number be derived from the current keymap?

File: libobs/obs-nix-wayland.c [hunk 4]
```diff
@@ -19,7 +19,224 @@
...
 
+#include <unistd.h>   // close
+#include <sys/mman.h> // mmap
+#include <string.h>   // strcmp
+
 #include <wayland-client.h>
+#include <xkbcommon/xkbcommon.h>
+
+// X11 only supports 256 scancodes, most keyboards dont have 256 keys so this should be reasonable.
+#define MAX_KEYCODES 256
```
Comment (GeorgesStavracas) [inline libobs/obs-nix-wayland.c:30]: Same for this one

File: libobs/obs-nix-wayland.c [hunk 5]
```diff
@@ -19,7 +19,224 @@
...
+static void platform_keyboard_keymap(void *data, struct wl_keyboard *keyboard,
+				     uint32_t format, int32_t fd, uint32_t size)
+{
+	UNUSED_PARAMETER(keyboard);
+	UNUSED_PARAMETER(format);
+	obs_hotkeys_platform_t *plat = (obs_hotkeys_platform_t *)data;
+
+	char *keymap_shm = mmap(NULL, size, PROT_READ, MAP_SHARED, fd, 0);
+	if (keymap_shm == MAP_FAILED) {
+		return;
```
Comment (GeorgesStavracas) [inline libobs/obs-nix-wayland.c:83]: I'm not sure if it's necessary to `close(fd)` here as well

File: libobs/obs-nix-wayland.c [hunk 6]
```diff
@@ -51,24 +281,1337 @@ obs_nix_wayland_hotkeys_platform_is_pressed(obs_hotkeys_platform_t *context,
...
+	/*
+	obs_hotkeys_platform_t *context = obs->hotkeys.platform_context;
+	struct keycode_list *keycodes = &context->keycodes[key];
+
+	for (size_t i = 0; i < keycodes->list.num; i++) {
+		if (get_key_translation(dstr, keycodes->list.array[i])) {
+			break;
+		}
+	}
+	*/
```
Comment (GeorgesStavracas) [inline libobs/obs-nix-wayland.c:393]: I'm not sure what were you intending to test here. Was it multi-layout?

File: libobs/obs-nix-wayland.c [hunk 7]
```diff
@@ -51,24 +281,1337 @@ obs_nix_wayland_hotkeys_platform_is_pressed(obs_hotkeys_platform_t *context,
...
+	obs_hotkeys_platform_t *plat = obs->hotkeys.platform_context;
+	// Translate the obs key back down to shift level 1 and then back to obs key.
+	xkb_keysym_t base_sym = plat->key_to_sym[plat->obs_to_key[key]][0];
```
Comment (GeorgesStavracas) [inline libobs/obs-nix-wayland.c:396]: Nitpick: could you please put `plat->obs_to_key[key]` in a separate variable?

**Review by GeorgesStavracas — CHANGES_REQUESTED** [review]
I left some comments and questions, but overall this looks pretty good to me. I didn't think of doing a roundtrip to get the wl_keyboard, what I had in mind is somehow get that from Qt, but that's probably a bad idea. What's here is an interesting approach.

File: libobs/obs-nix-wayland.c (hunk 6 above)
Comment (kkartaltepe) [reply libobs/obs-nix-wayland.c:393]: Oh sorry this was copied from the X11 impl, meant to delete.

File: libobs/obs-nix-wayland.c (hunk 3 above)
Comment (kkartaltepe) [reply libobs/obs-nix-wayland.c:32]: These could be derived but if we consider non-X11 compatible both keycodes and shift levels have no limit. So such a simple implementation may not be feasible. Only supporting X11 compatible keymaps seems like a nice middle ground of simplicity and supporting all existing (x11 compatible) layouts.

File: libobs/obs-nix-wayland.c [hunk 8]
```diff
@@ -19,7 +19,224 @@
...
+			     void *data)
+{
+	obs_hotkeys_platform_t *plat = (obs_hotkeys_platform_t *)data;
+	if (key >= MAX_KEYCODES)
+		return;
+
+	const xkb_keysym_t *syms;
+	// My setup doesnt have multiple layouts, do people actually use these?
+	// int num_layouts = xkb_keymap_num_layouts_for_key(keymap, key);
+	// blog(LOG_INFO, "key layouts: %d", num_layouts);
```
[the remark this answers is not in the document]
Comment (kkartaltepe) [reply libobs/obs-nix-wayland.c:57]: based on https://www.x.org/releases/X11R7.6/doc/xorg-docs/input/XKB-Config.html#id2520758 and an overview of the keyboard layouts preloaded on archlinux. There are no multi-layout keyboards defined by default, the only case for this would be users with custom defined xkb keymaps (so like 3 people?) where there appears to be like 4 possible layout levels.

I dont mind also hardcoding in layouts if we assume there can be up to 4 of them, though im not sure we really want to support arbitrary groups too, especially not if IMEs do change keymaps instead.

[the remark this answers is not in the document]
Comment (kkartaltepe) [reply libobs/obs-nix-wayland.c:57]: I played with some IME's and gnome, it looks like IMEs will use their own path for inputting characters for text input and do not use this mechanism at all.

For users configuring keyboards via gnome settings and not using IMEs gnome will only ever configure 4 layouts/groups at a time to remain X11 compatible. If you switch past the 4th keyboard gnome will reconfigure the wayland keyboard.

In my minimal playing this code seems to work fine for multiple languages once we handle multiple layouts/groups.

[the remark this answers is not in the document]
Comment (GeorgesStavracas) [reply libobs/obs-nix-wayland.c:57]: Alright, thanks for checking this

[the remark this answers is not in the document]
Comment (kkartaltepe) [reply libobs/obs-nix-wayland.c]: Pushed changes, I build the mapping tables any time users switch layouts. In cases where a user has something like a qwerty and azerty layout, scancode for AD01 might be bound to 'q' and 'a' which isnt quite correct. Switching keymaps takes 0.016ms on my machine so I think there is no problem with computing the lookup tables when users swap layouts to make the code simpler.

**Review by GeorgesStavracas — CHANGES_REQUESTED** [review]
The code looks good to me, I think it's good to go. However, I'd like to ask capitalizing `wayland` → `Wayland`

**Comment by kkartaltepe** [comment]
Updated, added some version checks when blinding `wl_seat`, checking the definition nothing important has changed for the past 8 years since v4 so we will bind whatever the compositor reports up to v7.

I think this is good to merge now.

**Review by GeorgesStavracas — APPROVED** [review]
